### ISSUE TITLE
Installation issue: py-jaxlib

### ISSUE THREAD

**Issue by omsai** [LLNL/spack#37548]
### Steps to reproduce the issue

```console
$ spack spec -I py-jaxlib                                                                                                              
Input spec                                                                                                                                                        
--------------------------------                                                                                                                                  
 -   py-jaxlib                                                                                                                                                    
                                                                                                                                                                  
Concretized                                                                                                                                                       
--------------------------------                                                                                                                                  
 -   py-jaxlib@0.4.3%gcc@11.2.0+cuda build_system=python_pip cuda_arch=none arch=linux-rocky8-zen3                                                                
[+]      ^bazel@5.4.0%gcc@11.2.0+nodepfail build_system=generic patches=0f6940d,3e6448a,53ca269,604423c,af73a49,dbbf38b arch=linux-rocky8-zen3
[+]          ^openjdk@11.0.17_8%gcc@11.2.0 build_system=generic arch=linux-rocky8-zen3                                                                            [+]          ^zip@3.0%gcc@11.2.0 build_system=makefile patches=14dc880,3bc30ba,5068e7c,51f48db,66ab4ce,a92fc4e,a95ed93,eb83fc8,f7d0bc4,fa8312c arch=linux-rocky8-z
en3                                                                                                                                                               
[+]      ^cuda@11.8.0%gcc@11.2.0~allow-unsupported-compilers~dev build_system=generic arch=linux-rocky8-zen3                                                      
[+]          ^libxml2@2.10.3%gcc@11.2.0~python build_system=autotools arch=linux-rocky8-zen3                                      
[+]              ^libiconv@1.17%gcc@11.2.0 build_system=autotools libs=shared,static arch=linux-rocky8-zen3                                                       
[+]      ^cudnn@8.4.0.27-11.6%gcc@11.2.0 build_system=generic arch=linux-rocky8-zen3                                                                              
[+]      ^py-numpy@1.24.2%gcc@11.2.0+blas+lapack build_system=python_pip patches=873745d arch=linux-rocky8-zen3                                                   
[+]          ^openblas@0.3.21%gcc@11.2.0~bignuma~consistent_fpcsr+fortran~ilp64+locking+pic+shared build_system=makefile patches=114f95f,d3d9b15 symbol_suffix=non
e threads=none arch=linux-rocky8-zen3                                                                                                                             
[+]              ^perl@5.36.0%gcc@11.2.0+cpanm+open+shared+threads build_system=generic arch=linux-rocky8-zen3
[+]                  ^berkeley-db@18.1.40%gcc@11.2.0+cxx~docs+stl build_system=autotools patches=26090f4,b231fcc arch=linux-rocky8-zen3
[+]          ^py-cython@0.29.32%gcc@11.2.0 build_system=python_pip arch=linux-rocky8-zen3                                                                         
[+]      ^py-pip@23.0%gcc@11.2.0 build_system=generic arch=linux-rocky8-zen3                                                                                      
[+]      ^py-scipy@1.10.1%gcc@11.2.0 build_system=python_pip arch=linux-rocky8-zen3                                                        
[+]          ^pkgconf@1.8.0%gcc@11.2.0 build_system=autotools arch=linux-rocky8-zen3                                                   
[+]          ^py-build@0.7.0%gcc@11.2.0~virtualenv build_system=python_pip patches=9a151ac arch=linux-rocky8-zen3
[+]              ^py-packaging@23.0%gcc@11.2.0 build_system=python_pip arch=linux-rocky8-zen3
[+]                  ^py-flit-core@3.7.1%gcc@11.2.0 build_system=python_pip arch=linux-rocky8-zen3
[+]              ^py-pep517@0.12.0%gcc@11.2.0 build_system=python_pip arch=linux-rocky8-zen3
[+]              ^py-tomli@2.0.1%gcc@11.2.0 build_system=python_pip arch=linux-rocky8-zen3
[+]          ^py-meson-python@0.12.0%gcc@11.2.0 build_system=python_pip arch=linux-rocky8-zen3
[+]              ^meson@1.0.0%gcc@11.2.0 build_system=python_pip patches=0f0b1bd arch=linux-rocky8-zen3
[+]              ^py-pyproject-metadata@0.6.1%gcc@11.2.0 build_system=python_pip arch=linux-rocky8-zen3
[+]          ^py-pybind11@2.10.1%gcc@11.2.0~ipo build_system=cmake build_type=RelWithDebInfo arch=linux-rocky8-zen3
[+]              ^cmake@3.25.2%gcc@11.2.0~doc+ncurses+ownlibs~qt build_system=generic build_type=Release arch=linux-rocky8-zen3
[+]              ^ninja@1.11.1%gcc@11.2.0+re2c build_system=generic arch=linux-rocky8-zen3
[+]                  ^re2c@2.2%gcc@11.2.0 build_system=generic arch=linux-rocky8-zen3
[+]          ^py-pythran@0.12.0%gcc@11.2.0 build_system=python_pip arch=linux-rocky8-zen3
[+]              ^py-beniget@0.4.1%gcc@11.2.0 build_system=python_pip arch=linux-rocky8-zen3
[+]              ^py-gast@0.5.3%gcc@11.2.0 build_system=python_pip arch=linux-rocky8-zen3
[+]              ^py-ply@3.11%gcc@11.2.0 build_system=python_pip arch=linux-rocky8-zen3
[+]      ^py-setuptools@63.0.0%gcc@11.2.0 build_system=generic arch=linux-rocky8-zen3
[+]      ^py-wheel@0.37.1%gcc@11.2.0 build_system=generic arch=linux-rocky8-zen3
[+]      ^python@3.10.8%gcc@11.2.0+bz2+crypt+ctypes+dbm~debug+libxml2+lzma~nis~optimizations+pic+pyexpat+pythoncmd+readline+shared+sqlite3+ssl~tkinter+uuid+zlib build_system=generic patches=0d98e93,7d40923,f2fd060 arch=linux-rocky8-zen3
[+]          ^bzip2@1.0.8%gcc@11.2.0~debug~pic+shared build_system=generic arch=linux-rocky8-zen3
[+]              ^diffutils@3.8%gcc@11.2.0 build_system=autotools arch=linux-rocky8-zen3
[+]          ^expat@2.5.0%gcc@11.2.0+libbsd build_system=autotools arch=linux-rocky8-zen3
[+]              ^libbsd@0.11.7%gcc@11.2.0 build_system=autotools arch=linux-rocky8-zen3
[+]                  ^libmd@1.0.4%gcc@11.2.0 build_system=autotools arch=linux-rocky8-zen3
[+]          ^gdbm@1.23%gcc@11.2.0 build_system=autotools arch=linux-rocky8-zen3
[+]          ^gettext@0.21.1%gcc@11.2.0+bzip2+curses+git~libunistring+libxml2+tar+xz build_system=autotools arch=linux-rocky8-zen3
[+]              ^tar@1.34%gcc@11.2.0 build_system=autotools zip=pigz arch=linux-rocky8-zen3
[+]                  ^pigz@2.7%gcc@11.2.0 build_system=makefile arch=linux-rocky8-zen3
[+]                  ^zstd@1.5.2%gcc@11.2.0+programs build_system=makefile compression=none libs=shared,static arch=linux-rocky8-zen3
[+]          ^libffi@3.4.4%gcc@11.2.0 build_system=autotools arch=linux-rocky8-zen3
[+]          ^libxcrypt@4.4.33%gcc@11.2.0~obsolete_api build_system=autotools arch=linux-rocky8-zen3
[+]          ^ncurses@6.4%gcc@11.2.0~symlinks+termlib abi=none build_system=autotools arch=linux-rocky8-zen3
[+]          ^openssl@1.1.1t%gcc@11.2.0~docs~shared build_system=generic certs=mozilla arch=linux-rocky8-zen3
[+]              ^ca-certificates-mozilla@2023-01-10%gcc@11.2.0 build_system=generic arch=linux-rocky8-zen3
[+]          ^readline@8.2%gcc@11.2.0 build_system=autotools patches=bbf97f1 arch=linux-rocky8-zen3
[+]          ^sqlite@3.40.1%gcc@11.2.0+column_metadata+dynamic_extensions+fts~functions+rtree build_system=autotools arch=linux-rocky8-zen3
[+]          ^util-linux-uuid@2.38.1%gcc@11.2.0 build_system=autotools arch=linux-rocky8-zen3
[+]          ^xz@5.4.1%gcc@11.2.0~pic build_system=autotools libs=shared,static arch=linux-rocky8-zen3
[+]          ^zlib@1.2.13%gcc@11.2.0+optimize+pic+shared build_system=makefile arch=linux-rocky8-zen3
```


### Error message

<details><summary>Error message</summary>

<pre>
==> Installing py-jaxlib-0.4.3-n2ofh47zprrm7gzdfmilrv7o5zaydzzr
==> No binary for py-jaxlib-0.4.3-n2ofh47zprrm7gzdfmilrv7o5zaydzzr found: installing from source
==> Using cached archive: /home/x-pnanda/src/spack/var/spack/cache/_source-cache/archive/21/2104735dc22be2b105e5517bd5bc6ae97f40e8e9e54928cac1585c6112a3d910.tar.gz
==> Ran patch() for py-jaxlib
==> py-jaxlib: Executing phase: 'install'
==> Error: ProcessError: Command exited with status 1:
    '/home/x-pnanda/src/spack/opt/spack/linux-rocky8-zen3/gcc-11.2.0/python-3.10.8-m2n5xokmokwkdi5zoorl6y553gpc62qs/bin/python3.10' 'build/build.py' '--bazel_startup_options=--output_user_root=/tmp/spackgoioj8ev'

3 errors found in build log:
     40    [4,806 / 6,186] Compiling tensorflow/compiler/xla/service/cpu/runtim
           e_matmul.cc; 157s local ... (128 actions, 125 running)
     41    [5,281 / 6,210] Compiling tensorflow/compiler/xla/service/cpu/runtim
           e_matmul.cc; 208s local ... (128 actions, 127 running)
     42    [5,579 / 6,210] Compiling tensorflow/compiler/xla/service/cpu/runtim
           e_matmul.cc; 267s local ... (128 actions, 127 running)
     43    [6,064 / 6,343] Compiling tensorflow/compiler/xla/service/cpu/runtim
           e_matmul.cc; 335s local ... (128 actions, 127 running)
     44    [6,179 / 6,343] Compiling mlir/lib/Dialect/LLVMIR/IR/LLVMDialect.cpp
           ; 449s local ... (128 actions running)
     45    [6,179 / 6,343] Compiling mlir/lib/Dialect/LLVMIR/IR/LLVMDialect.cpp
           ; 533s local ... (128 actions running)
  >> 46    ERROR: /tmp/spackgoioj8ev/b25779cec22498dc331845aa03eb6233/external/
           org_tensorflow/tensorflow/compiler/xla/mlir_hlo/BUILD:434:11: Compil
           ing tensorflow/compiler/xla/mlir_hlo/mhlo/IR/hlo_ops.cc failed: (Exi
           t 1): gcc failed: error executing command
     47      (cd /tmp/spackgoioj8ev/b25779cec22498dc331845aa03eb6233/execroot/_
           _main__ && \
     48      exec env - \
     49        PATH=/home/x-pnanda/src/spack/opt/spack/linux-rocky8-zen3/gcc-11
           .2.0/py-numpy-1.24.2-bcjinwartqsaojq6sjei3x2b5gmgbmji/bin:/home/x-pn
           anda/src/spack/opt/spack/linux-rocky8-zen3/gcc-11.2.0/py-wheel-0.37.
           1-czdncuoyrpmsqlfoa5r6bvstxnl6mloc/bin:/home/x-pnanda/src/spack/opt/
           spack/linux-rocky8-zen3/gcc-11.2.0/py-pip-23.0-niv56i3wdsioegnxlh63b
           sshk6yuhsym/bin:/home/x-pnanda/src/spack/opt/spack/linux-rocky8-zen3
           /gcc-11.2.0/bazel-5.4.0-6g2kczo2kqotaeat2pwt6e5fcnpyn5d4/bin:/home/x
           -pnanda/src/spack/opt/spack/linux-rocky8-zen3/gcc-11.2.0/zip-3.0-ksq
           2oxktcw4qcxy6cmznizdcl56fcfj3/bin:/home/x-pnanda/src/spack/opt/spack
           /linux-rocky8-zen3/gcc-11.2.0/python-3.10.8-m2n5xokmokwkdi5zoorl6y55
           3gpc62qs/bin:/home/x-pnanda/src/spack/opt/spack/linux-rocky8-zen3/gc
           c-11.2.0/python-3.10.8-m2n5xokmokwkdi5zoorl6y553gpc62qs/bin:/home/x-
           pnanda/src/spack/opt/spack/linux-rocky8-zen3/gcc-11.2.0/openjdk-11.0
           .17_8-wgyvgctbgqzf7ct4oslha2wcylvkyjit/bin:/home/x-pnanda/src/spack/
           lib/spack/env/gcc:/home/x-pnanda/src/spack/lib/spack/env/case-insens
           itive:/home/x-pnanda/src/spack/lib/spack/env:/home/x-pnanda/src/spac
           k/var/spack/environments/abc/.spack-env/view/bin:/home/x-pnanda/src/
           spack/var/spack/environments/login/.spack-env/view/bin:/apps/spack/a
           nvil/apps/fftw/3.3.8-gcc-11.2.0-wmcvzic/bin:/apps/spack/anvil/apps/g
           sl/2.4-gcc-8.4.1-exxsoc2/bin:/apps/spack/anvil/apps/openmpi/4.0.6-gc
           c-11.2.0-3navcwb/bin:/apps/spack/anvil/apps/numactl/2.0.14-gcc-11.2.
           0-wrjotmv/bin:/apps/spack/anvil/apps/gcc/11.2.0-gcc-8.4.1-qjtdkvs/bi
           n:/home/x-pnanda/src/spack/bin:/apps/spack/anvil/apps/libfabric/1.12
           .0-gcc-8.4.1-xj6lmd4/bin:/home/x-pnanda/bin:/home/x-pnanda/.local/bi
           n:/usr/local/bin:/usr/bin:/usr/local/sbin:/usr/sbin:/opt/puppetlabs/
           bin:/usr/site/rcac/sbin:/usr/site/rcac/bin:/usr/site/rcac/scripts:/o
           pt/thinlinc/bin:/home/x-pnanda/immunology/scripts:/apps/spack/anvil/
           apps/matlab/R2020b-gcc-8.4.1-ybog64f/bin/glnxa64:/apps/spack/anvil/a
           pps/matlab/R2020b-gcc-8.4.1-ybog64f/bin \
     50        PWD=/proc/self/cwd \
     51        SPACK_CC=/apps/spack/anvil/apps/gcc/11.2.0-gcc-8.4.1-qjtdkvs/bin
           /gcc \
     52        SPACK_CC_RPATH_ARG=-Wl,-rpath, \

     ...

     81    # Execution platform: @local_execution_config_platform//:platform
     82    gcc: fatal error: Killed signal terminated program cc1plus
     83    compilation terminated.
     84    Target //build:build_wheel failed to build
     85    INFO: Elapsed time: 900.298s, Critical Path: 604.71s
     86    INFO: 6307 processes: 890 internal, 5417 local.
  >> 87    FAILED: Build did NOT complete successfully
     88    ERROR: Build failed. Not running target
  >> 89    FAILED: Build did NOT complete successfully
     90    
     91         _   _  __  __
     92        | | / \ \ \/ /
     93     _  | |/ _ \ \  /
     94    | |_| / ___ \/  \
     95     \___/_/   \/_/\_\

See build log for details:
  /anvil/scratch/x-pnanda/spack-stage/spack-stage-py-jaxlib-0.4.3-n2ofh47zprrm7gzdfmilrv7o5zaydzzr/spack-build-out.txt

==> Warning: Skipping build of py-pytensor-2.10.1-pm57wqyc3trvxpiu42olmd65jbfqdz3v since py-jaxlib-0.4.3-n2ofh47zprrm7gzdfmilrv7o5zaydzzr failed
==> Warning: Skipping build of py-pymc-5.0.2-yr6tueppnn62v6f5az7c7dj2exwn22dh since py-pytensor-2.10.1-pm57wqyc3trvxpiu42olmd65jbfqdz3v failed
==> Warning: Skipping build of py-jax-0.4.3-4tccmofcx4zynxmf2scmvhtwkdgkcfll since py-jaxlib-0.4.3-n2ofh47zprrm7gzdfmilrv7o5zaydzzr failed
==> Error: py-pymc-5.0.2-yr6tueppnn62v6f5az7c7dj2exwn22dh: Package was not installed
==> Error: Installation request failed.  Refer to reported errors for failing package(s).
[+] /home/x-pnanda/src/spack/opt/spack/linux-rocky8-zen3/gcc-11.2.0/py-matplotlib-3.7.0-rmnhjyhkw4g3r3apo5pkr2paq62ww2h4
[+] /home/x-pnanda/src/spack/opt/spack/linux-rocky8-zen3/gcc-11.2.0/py-xarray-2022.9.0-jajmz74r3wks6ff62uq54ixpwqxdj7in
[+] /home/x-pnanda/src/spack/opt/spack/linux-rocky8-zen3/gcc-11.2.0/py-minikanren-1.0.3-3uas3uwhztxytw3xyjtmdutdl6mjnq7u
[+] /home/x-pnanda/src/spack/opt/spack/linux-rocky8-zen3/gcc-11.2.0/py-xarray-einstats-0.5.1-23ob4p6q42wufcfewje6tcyd6r3sb3j2
[+] /home/x-pnanda/src/spack/opt/spack/linux-rocky8-zen3/gcc-11.2.0/py-arviz-0.15.0-qfaiv7euswgqi4q5pqsfggx47jq6rev4
slurmstepd: error: Detected 1 oom-kill event(s) in StepId=1513174.batch. Some of your processes may have been killed by the cgroup out-of-memory handler.
</pre></details>


### Information on your system

* **Spack:** 0.20.0.dev0 (9c458751551aa8e2bf204a5b1abbd021472e59c7)
* **Python:** 3.6.8
* **Platform:** linux-rocky8-zen3
* **Concretizer:** clingo

### Additional information

@adamjstewart  @pradyunsg

So the issue is `build_wheel` of jaxlib on the cluster uses all the 128 CPU cores instead of those the smaller proportion of CPUs assigned by cgroups via SLURM, causing the build to run out of the cgroups allocated memory (memory is 1.9 GB / CPU core). I finally was able to get the library to compile by allocating 160 GB.

In `spack-build-out.txt` you can see the `[... / 6,431] Compiling ...` lines using 128 actions / CPU cores.  I do see the patch to use `--jobs={make_jobs}`, but that only seems to apply to installing the wheel rather than during compilation.

These are the output files after increasing the memory showing the discrepancy between actions and `--jobs` when it successfully builds: 

* [spack-build-out.txt](https://github.com/spack/spack/files/11416099/spack-build-out.txt)
* [spack-build-env.txt](https://github.com/spack/spack/files/11416100/spack-build-env.txt)

### General information

- [X] I have run `spack debug report` and reported the version of Spack/Python/Platform
- [X] I have run `spack maintainers <name-of-the-package>` and **@mentioned** any maintainers
- [X] I have uploaded the build log and environment files
- [X] I have searched the issues of this repo and believe this is not a duplicate

**Comment by adamjstewart** [comment]
Thanks for tracking this down, I was having the same issue on our 128 core node. The only way I could get jax to build was to give it all 512 GB of RAM.

> I do see the patch to use `--jobs={make_jobs}`, but that only seems to apply to installing the wheel rather than during compilation.

Any idea where the correct place to patch would be? We should also check other packages too. I tried fixing this in #36002 but clearly didn't do a good job.

**Comment by omsai** [comment]
At the beginning of the [`spack-build-out.txt`](https://github.com/spack/spack/files/11416099/spack-build-out.txt) log the Bazel server reads the `.bazelrc` file at the root of the `py-jaxlib-*/spack-src` stage directory as well as the `.jax_configure.bazelrc` file written by `build/build.py`. I think maybe the thing to do is patch one of the bazelrc files. SLURM uses cgroups to control the number of CPUs and RAM allocated; Bazel supports limiting memory with cgroups via [--experimental_sandbox_memort_limit_mb](https://bazel.build/reference/command-line-reference#flag--experimental_sandbox_memory_limit_mb). For CPUs, there are several options when one searches for "HOST_CPUS" in the command-line reference:
1. [--loading_phase_threads](https://bazel.build/reference/command-line-reference#flag--loading_phase_threads)
2. [--worker_max_instances](https://bazel.build/reference/command-line-reference#flag--worker_max_instances) which needs the 'worker' strategy enabled.
3. [--worker_max_multiplex_instances](https://bazel.build/reference/command-line-reference#flag--worker_max_multiplex_instances) which needs the 'worker' strategy enabled.
4. [--local_test_jobs](https://bazel.build/reference/command-line-reference#flag--local_test_jobs)
5. [--local_cpu_resources](https://bazel.build/reference/command-line-reference#flag--local_cpu_resources)

**Comment by adamjstewart** [comment]
I know next to nothing about bazel but if you want to submit a PR I would be willing to test it. @aweits @glennpj have more experience patching bazel than I do, maybe they can comment on which flag is best.

**Comment by omsai** [comment]
Solved it! It was the last option in the list above. One needs to add the Bazel server option `--local_cpu_resources={make_jobs}` in the top-level `.bazelrc` after the `--spawn_strategy=local` as hinted by the note in `--jobs`: https://bazel.build/docs/user-manual#jobs and the compilation now runs using the new server-side CPU limit without trying to use all the HOST_CPUS of the entire SLURM compute node. PR incoming.

**Comment by omsai** [comment]
Closing because merged in #38090
